keps/sig-auth/3766-referencegrant/README.md

```diff
@@ -441,174 +368,365 @@ rejected".
+  // +kubebuilder:validation:MaxItems=16
+  Resources []GroupResource `json:"resources"`
+
+  // ReferenceGrantNamespace describes the namespace(s) of the referent.
+  ReferenceGrantNamespace string `json:"namespace"`
 }
 
-```
-<<[UNRESOLVED Are "To" and "From" the right field names? ]>>
-Previous Discussion: https://github.com/kubernetes/enhancements/pull/3767#discussion_r1084671720
+// GroupResource describes trusted groups and resources.
+type GroupResource struct {
+  // Group is the group of the referent.
+  // When empty, the Kubernetes core API group is inferred.
+  Group Group `json:"group"`
+
+  // Resource is the resource of the referent.
+  Resource string `json:"resource"`
+}
 
-Comments from @thockin:
+// Namespace describes trusted namespaces. Exactly one of Name or Selector MUST be specified.
+// +oneOf
+type ReferenceGrantNamespace struct {
+  // Name is the name of the namespace.
+  Name string `json:"name"`
 
-I would not argue for subject/object - that's confusing, and I love to take
-analogies too far.
+  // Selector selects namespaces based on their labels.
+  Selector metav1.LabelSelector `json:"selector"`
+}
+
+
+// ReferenceGrantTo describes what resources are allowed as targets of the
+// references.
+type ReferenceGrantTo struct {
+  // Group is the group of the referent.
+  // When empty, the Kubernetes core API group is inferred.
+  Group string `json:"group"`
+
+  // Resource is the resource of the referent.
+  Resource string `json:"resource"`
+
+  // Name is the name of the referent. When unspecified, this policy
+  // refers to all resources of the specified Group and Kind in the local
+  // namespace.
+  //
+  // +optional
+  Name *string `json:"name,omitempty"`
+}
+```
 
-NetPol uses To/From but only for actual communications, and that still has been
-critiqued as perhaps "too cute".
+### Variations from Gateway API ReferenceGrant
 
-Subject/From isn't too bad, but not as symmetric. Subject/Referrer is correct
-but decidedly uncute. Subject/Origin?
+This KEP makes several changes from the existing ReferenceGrant resource
+defined by Gateway API:
 
-I hold opinions from an API review POV, but I'd like sig-auth to own the
-decision :)
+#### 1. Removing Space for Status
 
-For reference, there was an [earlier
-discussion](https://groups.google.com/g/kubernetes-api-reviewers/c/ldmrXXQC4G4)
-on the kubernetes-api-reviewers mailing list that's also relevant to this.
-<<[/UNRESOLVED]>>
+Although we had theoretical use cases for status, they never materialized into
+the Gateway API version of ReferenceGrant. [Earlier feedback on this
+KEP](https://github.com/kubernetes/enhancements/pull/3767#discussion_r1084670421)
+showed that we could use alternative approaches such as Events or new
+"RefGranted" resources published by implementing controllers. Given that there
+are both reasonable alternatives to status and we have not needed it so far, it
+seems reasonable to leave out of this version of the API.
```

Comment:
I'd prefer ReferenceGrant to be an object and to follow our conventions for objects, which include:
- optional `status` field at top level
- **one** specification field at top level (my emphasis)

We don't need to call the top level specification field `spec` (another option could be `grant`). However, if we one day want to add fields such as `notBeforeDate`, `notAfterDate`, `enabled`, `conditionExpression` etc - we will not have left ourselves as much oom for that. Having lots of top level fields is quite a bit less ergonomic.

For this reason, I recommend that we leave the `.spec` field at the top level.